sensing/autoware_pointcloud_preprocessor/src/distortion_corrector/distortion_corrector.cpp

```diff
@@ -187,15 +175,139 @@ bool DistortionCorrector<T>::isInputValid(sensor_msgs::msg::PointCloud2 & pointc
+        ? abs(next_cartesian_rad - current_cartesian_rad) - 2 * autoware::universe_utils::pi
+        : next_cartesian_rad - current_cartesian_rad;
+
+    float sign = azimuth_diff / cartesian_rad_diff;
+
+    // Check if 'sign' can be adjusted to 1 or -1
+    if (std::abs(sign - 1.0f) <= angle_conversion.sign_threshold) {
+      angle_conversion.sign = 1.0f;
+    } else if (std::abs(sign + 1.0f) <= angle_conversion.sign_threshold) {
+      angle_conversion.sign = -1.0f;
+    } else {
+      RCLCPP_DEBUG_STREAM_THROTTLE(
+        node_->get_logger(), *node_->get_clock(), 10000 /* ms */,
+        "Value of sign is not close to 1 or -1. Iterate to next point ...");
+      continue;
+    }
+
+    float offset_rad = *it_azimuth - sign * current_cartesian_rad;
+    // Check if 'offset_rad' can be adjusted to offset_rad multiple of π/2
+    int multiple_of_90_degrees = std::round(offset_rad / (autoware::universe_utils::pi / 2));
+    if (
+      std::abs(offset_rad - multiple_of_90_degrees * (autoware::universe_utils::pi / 2)) >
+      angle_conversion.offset_rad_threshold) {
+      RCLCPP_DEBUG_STREAM_THROTTLE(
+        node_->get_logger(), *node_->get_clock(), 10000 /* ms */,
+        "Value of offset_rad is not close to multiplication of 90 degrees. Iterate to next point "
+        "...");
+      continue;
+    }
+
+    // Limit the range of offset_rad in [0, 360)
+    multiple_of_90_degrees = (multiple_of_90_degrees % 4 + 4) % 4;
+
+    angle_conversion.offset_rad = multiple_of_90_degrees * (autoware::universe_utils::pi / 2);
+
+    return angle_conversion;
+  }
+  return std::nullopt;
+}
+
+void DistortionCorrectorBase::warn_if_timestamp_is_too_late(
+  bool is_twist_time_stamp_too_late, bool is_imu_time_stamp_too_late)
+{
+  if (is_twist_time_stamp_too_late) {
+    RCLCPP_WARN_STREAM_THROTTLE(
+      node_->get_logger(), *node_->get_clock(), 10000 /* ms */,
+      "Twist time_stamp is too late. Could not interpolate.");
+  }
+
+  if (is_imu_time_stamp_too_late) {
+    RCLCPP_WARN_STREAM_THROTTLE(
+      node_->get_logger(), *node_->get_clock(), 10000 /* ms */,
+      "IMU time_stamp is too late. Could not interpolate.");
+  }
+}
+
+void DistortionCorrectorBase::convert_matrix_to_transform(
+  const Eigen::Matrix4f & matrix, tf2::Transform & transform)
+{
+  transform.setOrigin(tf2::Vector3(matrix(0, 3), matrix(1, 3), matrix(2, 3)));
+  transform.setBasis(tf2::Matrix3x3(
+    matrix(0, 0), matrix(0, 1), matrix(0, 2), matrix(1, 0), matrix(1, 1), matrix(1, 2),
+    matrix(2, 0), matrix(2, 1), matrix(2, 2)));
+}
+
 template <class T>
-void DistortionCorrector<T>::undistortPointCloud(
-  bool use_imu, sensor_msgs::msg::PointCloud2 & pointcloud)
+void DistortionCorrector<T>::undistort_pointcloud(
+  bool use_imu, std::optional<AngleConversion> angle_conversion_opt,
+  sensor_msgs::msg::PointCloud2 & pointcloud)
 {
-  if (!isInputValid(pointcloud)) return;
+  if (!is_pointcloud_valid(pointcloud)) return;
+  if (twist_queue_.empty()) {
```

Comment:
It seems you got the twist check outside the isInputValid because you needed a version for the pointcloud.
In that case you could have done them independently while also keeping the isInputValid

Reply:
I think having both `is_input_valid` and `is_pointlcoud_valid` would be a bit confusing.
And in the `is_input_valid` there is only an additional check (3 lines) for checking the twist queue. Also, there is only one place that will call this function.